### PR TITLE
[IVY] fix node injection

### PR DESCRIPTION
**Pull request by vicb**

### CODE REVIEW

File: packages/core/src/render3/di.ts [hunk 1]
```diff
@@ -433,16 +423,15 @@ function searchMatchesQueuedForCreation<T>(node: LNode, token: any): T|null {
...
- * ID must be modulo-ed by BLOOM_SIZE to get the correct bloom bit (directives share slots after
- * BLOOM_SIZE is reached).
+ * When a directive is public, it is added to the bloom filter and given a unique ID that can be
+ * retrieved on the Type. When the directive isn't public or the token is not a directive `null`
+ * is returned as the node injector can not possibly provide that token.
  *
- * @param type The directive type
- * @returns The bloom bit to check for the directive
+ * @param token the injection token
+ * @returns the matching bit to check in the bloom filter of null if the token is not known.
```
Comment (JoostK) [inline packages/core/src/render3/di.ts]: `of null` > `or null`

File: packages/core/src/render3/di.ts [hunk 2]
```diff
@@ -177,10 +178,12 @@ export function diPublic(def: DirectiveDefInternal<any>): void {
...
  * @returns the value from the injector or `null` when not found
  */
-export function directiveInject<T>(token: Type<T>): T;
-export function directiveInject<T>(token: Type<T>, flags: InjectFlags.Optional): T|null;
-export function directiveInject<T>(token: Type<T>, flags: InjectFlags): T;
-export function directiveInject<T>(token: Type<T>, flags = InjectFlags.Default): T|null {
+export function directiveInject<T>(token: Type<T>| InjectionToken<T>): T;
+export function directiveInject<T>(
+    token: Type<T>| InjectionToken<T>, flags: InjectFlags.Optional): T|null;
+export function directiveInject<T>(token: Type<T>| InjectionToken<T>, flags: InjectFlags): T;
```
Comment (JoostK) [inline packages/core/src/render3/di.ts]: This signature is inaccurate (and used to be before this PR), as it depends on the exact flag whether or not the return type should include `null`.

Comment (gkalpak) [reply packages/core/src/render3/di.ts]: Are there other values (apart from `InjectFlags.Optional`) that can result in `null`?

File: packages/core/src/render3/di.ts [hunk 3]
```diff
@@ -174,17 +168,16 @@ export function diPublic(def: DirectiveDefInternal<any>): void {
...
+ * @returns the value from the injector or `null` when not found
  */
-export function directiveInject<T>(token: Type<T>): T;
-export function directiveInject<T>(token: Type<T>, flags: InjectFlags.Optional): T|null;
-export function directiveInject<T>(token: Type<T>, flags: InjectFlags): T;
-export function directiveInject<T>(token: Type<T>, flags = InjectFlags.Default): T|null {
+export function directiveInject<T>(token: Type<T>| InjectionToken<T>): T;
+export function directiveInject<T>(
+    token: Type<T>| InjectionToken<T>, flags: InjectFlags.Optional): T|null;
+export function directiveInject<T>(token: Type<T>| InjectionToken<T>, flags: InjectFlags): T|null;
```
Comment (gkalpak) [inline packages/core/src/render3/di.ts]: I think this signature is wrong :grin:
AFAICT, the intention is to have signatures for specific types (that appear above this one) that can produce `null` and have this signature to catch all the rest that cannot produce `null`.

If there are other `InjectFlags` values that can produce `null`, then they should be part of the signature above.
But this signature should not have `null` in the return value.

File: packages/core/src/render3/di.ts (hunk 2 above)
Comment (gkalpak) [reply packages/core/src/render3/di.ts]: No worries. Can happen to anyone (even @vicb 😛).

File: packages/core/src/render3/di.ts [hunk 4]
```diff
@@ -61,7 +62,7 @@ let nextNgElementId = 0;
  * @param type The directive to register
  */
 export function bloomAdd(injector: LInjector, type: Type<any>): void {
-  let id: number|undefined = (type as any)[NG_ELEMENT_ID];
+  let id = (type as any)[NG_ELEMENT_ID];
```
Comment (mhevery) [inline packages/core/src/render3/di.ts]: Why did you remove `: number|undefined`. Could we add it back?

File: packages/core/src/render3/di.ts (hunk 3 above)
Comment (vicb) [reply packages/core/src/render3/di.ts]: I think this line can indeed be removed.
_I don't think there is any way to express what you say using the type system given we use flags_

File: packages/core/src/render3/di.ts [hunk 5]
```diff
@@ -61,7 +62,7 @@ let nextNgElementId = 0;
...
 
   // Set a unique ID on the directive type, so if something tries to inject the directive,
   // we can easily retrieve the ID and hash it into the bloom bit that should be checked.
```
Comment (mhevery) [inline packages/core/src/render3/di.ts]: since you are already here could we rewrite the code to use bit operations rather than `%` and `<` operations?

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
Could we add tests to `render3/di_spec.ts` for this?

File: packages/core/src/render3/di.ts (hunk 3 above)
Comment (JoostK) [reply packages/core/src/render3/di.ts]: If you remove it fully, it won't accept flags other than `InjectFlags.Optional` at all. [See playground example](https://www.typescriptlang.org/play/index.html#src=export%20const%20enum%20InjectFlags%20%7B%0D%0A%20%20Default%20%3D%200%2C%0D%0A%20%20Optional%20%3D%201%2C%0D%0A%7D%0D%0A%0D%0Aexport%20type%20Type%3CT%3E%20%3D%20Function%3B%0D%0A%0D%0Aexport%20function%20directiveInject%3CT%3E(token%3A%20Type%3CT%3E)%3A%20T%3B%0D%0Aexport%20function%20directiveInject%3CT%3E(token%3A%20Type%3CT%3E%2C%20flags%3A%20InjectFlags.Optional)%3A%20T%7Cnull%3B%0D%0A%2F%2F%20export%20function%20directiveInject%3CT%3E(token%3A%20Type%3CT%3E%2C%20flags%3A%20InjectFlags)%3A%20T%3B%0D%0Aexport%20function%20directiveInject%3CT%3E(token%3A%20Type%3CT%3E%2C%20flags%20%3D%20InjectFlags.Default)%3A%20T%20%7C%20null%20%7B%0D%0A%20%20return%20null%3B%0D%0A%7D%0D%0A%0D%0AdirectiveInject(Array%2C%20InjectFlags.Default)%3B%0D%0A).

Comment (vicb) [reply packages/core/src/render3/di.ts]:
```ts
export function directiveInject<T>(token: Type<T>| InjectionToken<T>): T;
export function directiveInject<T>(
    token: Type<T>| InjectionToken<T>, flags: InjectFlags = InjectFlags.Default): T|null {
  return getOrCreateInjectable<T>(getOrCreateNodeInjector(), token, flags);
}
```

Comment (gkalpak) [reply packages/core/src/render3/di.ts]: Yes, the line should not be removed. It needs to be chanfged to return `T` (not `T|null`).
The rest is fine.

File: packages/core/src/render3/di.ts [hunk 6]
```diff
@@ -174,17 +170,14 @@ export function diPublic(def: DirectiveDefInternal<any>): void {
...
- *
- * @param token The directive type to search for
- * @param flags Injection flags (e.g. CheckParent)
- * @returns The instance found
+ * @param token the type or token to inject
+ * @param flags Injection flags
+ * @returns the value from the injector or `null` when not found
  */
-export function directiveInject<T>(token: Type<T>): T;
-export function directiveInject<T>(token: Type<T>, flags: InjectFlags.Optional): T|null;
```
Comment (gkalpak) [inline packages/core/src/render3/di.ts]: No, no no. This line should stay :grin:

Comment (gkalpak) [reply packages/core/src/render3/di.ts]: I.e. this function should be defined as:

```ts
export function directiveInject<T>(token: Type<T>| InjectionToken<T>): T;
export function directiveInject<T>(token: Type<T>| InjectionToken<T>, flags: InjectFlags.Optional): T|null;
export function directiveInject<T>(token: Type<T>| InjectionToken<T>, flags: InjectFlags): T;
export function directiveInject<T>(
    token: Type<T>| InjectionToken<T>, flags = InjectFlags.Default): T|null {
  return getOrCreateInjectable<T>(getOrCreateNodeInjector(), token, flags);
}
```

**Review by mhevery — APPROVED** [review]

File: packages/core/src/render3/di.ts (hunk 6 above)
Comment (vicb) [reply packages/core/src/render3/di.ts]: and what if flags = `InjectFlags.Optional | InjectFlags.SkipSelf` ?

**Comment by vicb** [comment]
http://test/OCL:206380401:BASE:206380532:1532729215132:194256de

**merge-assistance**: [golden symbol changes](https://github.com/angular/angular/pull/25166/files#diff-8bec588d7a22fe9ab8f11a8114e80414) are only some renaming - no impact on size.

**Review by mhevery — APPROVED** [review]
